x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+	} else {
+		suite.Require().Contains(string(ack), "error",
+			"acknoledgment is not an error")
+		suite.Require().Contains(string(ack), types.RateLimitExceededMsg,
+			"acknoledgment error is not of the right type")
+	}
+	return ack, err
+}
+
+func (suite *MiddlewareTestSuite) AssertSendSuccess(success bool, msg sdk.Msg) (*sdk.Result, error) {
+	r, err := suite.chainA.SendMsgsNoCheck(msg)
+	if success {
+		suite.Require().NoError(err, "IBC send failed. Expected success. %s", err)
+	} else {
+		suite.Require().Error(err, "IBC send succeeded. Expected failure")
+		suite.ErrorContains(err, types.RateLimitExceededMsg, "Bad error type")
+	}
+	return r, err
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuota(name string, duration, send_precentage, recv_percentage uint32) string {
+	return fmt.Sprintf(`
+          {"channel_id": "channel-0", "denom": "%s", "quotas": [{"name":"%s", "duration": %d, "send_recv":[%d, %d]}] }
+    `, sdk.DefaultBondDenom, name, duration, send_precentage, recv_percentage)
+}
+
+// Tests
+
+// Test that Sending IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestSendTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, one))
+}
+
+// Test that Receiving IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestReceiveTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertReceiveSuccess(true, suite.NewValidMessage(false, one))
+}
+
+// Test rate limiting on sends
+func (suite *MiddlewareTestSuite) TestSendTransferWithRateLimiting() map[string]string {
+	// Setup contract
+	suite.chainA.StoreContractCode(&suite.Suite)
+	quotas := suite.BuildChannelQuota("weekly", 604800, 5, 5)
+	addr := suite.chainA.InstantiateContract(&suite.Suite, quotas)
+	suite.chainA.RegisterRateLimitingContract(addr)
+
+	// Setup sender chain's quota
+	osmosisApp := suite.chainA.GetOsmosisApp()
+
+	// Each user has 10% of the supply
+	supply := osmosisApp.BankKeeper.GetSupply(suite.chainA.GetContext(), sdk.DefaultBondDenom)
+	quota := supply.Amount.QuoRaw(20)
+	half := quota.QuoRaw(2)
+
+	// send 2.5% (quota is 5%)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
+
+	// send 2.5% (quota is 5%)
+	r, _ := suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
+
+	// Calculate remaining allowance in the quota
+	attrs := suite.ExtractAttributes(suite.FindEvent(r.GetEvents(), "wasm"))
+	used, _ := sdk.NewIntFromString(attrs["weekly_used_out"])
+	suite.Require().Equal(used, half.MulRaw(2))
+
+	// Sending above the quota should fail.
+	suite.AssertSendSuccess(false, suite.NewValidMessage(true, sdk.NewInt(1)))
+	return attrs
+}
+
+// Test rate limits are reset when the specified time period has passed
+func (suite *MiddlewareTestSuite) TestSendTransferReset() {
+	// Same test as above, but the quotas get reset after time passes
+	attrs := suite.TestSendTransferWithRateLimiting()
```

Comment:
Maybe I'm just unused to this but I've never seen returning results from a test spec. Is it possible to avoid this for consistency's sake?

Reply:
yeah, it's a bit weird. Just did it to avoid either repeating the code or abstracting it into a function that's only used twice. If we abstract it, the helper function would basically be exactly what `TestSendTransferWithRateLimiting` is now, and the test just a call to that function. 

Not sure which is better, but happy to change it. Do you have a better alternative?